src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs

```diff
@@ -84,5 +224,187 @@ private protected IReadOnlyList<PackageReference> GetTransitivePackageReferences
+        private async ValueTask<IList<LockFileTarget>> GetTargetsListAsync(string assetsFilePath, CancellationToken ct)
+        {
+            ct.ThrowIfCancellationRequested();
+
+            await TaskScheduler.Default;
+
+            LockFile lockFile = LockFileUtilities.GetLockFile(assetsFilePath, NullLogger.Instance);
+
+            return lockFile?.Targets;
+        }
+
+        /// <summary>
+        /// Runs Depth First Search recursively to mark current and dependend nodes with top dependency
+        /// </summary>
+        /// <param name="top">Top, Direct dependency</param>
+        /// <param name="current">Current package/node to visit</param>
+        /// <param name="graph">Package dependency graph, from assets file</param>
+        /// <param name="visited">Dictionary to remember visited nodes</param>
+        /// <param name="fxRidEntry">Framework/Runtime-ID associated with current <paramref name="graph"/></param>
+        private void MarkTransitiveOrigin(PackageReference top, PackageReference current, LockFileTarget graph, HashSet<PackageIdentity> visited, FrameworkRuntimePair fxRidEntry, CancellationToken token)
+        {
+            token.ThrowIfCancellationRequested();
+
+            LockFileTargetLibrary node = default;
+
+            // Find first target node that matches current
+            foreach (LockFileTargetLibrary lib in graph.Libraries)
+            {
+                if (lib.Type == LibraryType.Package.Value
+                    && string.Equals(lib.Name, current.PackageIdentity.Id, StringComparison.OrdinalIgnoreCase))
+                {
+                    node = lib;
+                    break;
+                }
+            }
+
+            /* 
+             * && ((current.HasAllowedVersions && current.AllowedVersions.Satisfies(lib.Version)) ||
+                        (current.PackageIdentity.HasVersion && current.PackageIdentity.Version.Equals(lib.Version)))
+             */
+
+            if (node != default)
+            {
+                visited.Add(current.PackageIdentity); // visited
+
+                // Lookup Transitive Origins Cache
+                TransitiveEntry cachedEntry;
+                if (!TransitiveOriginsCache.TryGetValue(current.PackageIdentity.Id, out cachedEntry))
+                {
+                    cachedEntry = new Dictionary<FrameworkRuntimePair, IList<PackageReference>>
+                    {
+                        [fxRidEntry] = new List<PackageReference>()
+                    };
+                }
+
+                if (!cachedEntry.ContainsKey(fxRidEntry))
+                {
+                    cachedEntry[fxRidEntry] = new List<PackageReference>();
+                }
+
+                if (!cachedEntry[fxRidEntry].Contains(top))
+                {
+                    cachedEntry[fxRidEntry].Add(top);
+                }
+
+                // Upsert Transitive Origins Cache
+                TransitiveOriginsCache[current.PackageIdentity.Id] = cachedEntry;
```

Comment:
If 2 versions of same package Id (i.e newsoft.json) may have different origin path tree. Are we accounting that here?

Reply:
It's better to index by only Package ID. Two packages could have different version in their dependencies, but, In PackageReference model, only one package ID will be resolved. For example:

ProjectA deps: 
```
<PackageReference Include="Pkg1" Version="1.0.0" />
<PackageReference Include="Pkg2" Version="2.0.1" />
```
ProjectB deps: Pkg1_3.0.0
```
<PackageReference Include="Pkg1" Version="3.0.0" />
```

Assets file will contain:
- ProjectA
- ProjectB
- Pkg1, version 3.0.0
- Pkg2, version 2.0.1